test/bufconn/bufconn_test.go

```diff
@@ -197,3 +197,103 @@ func TestCloseWhileAccepting(t *testing.T) {
 		t.Fatalf("c, err = %v, %v; want nil, %v", c, err, errClosed)
 	}
 }
+
+func TestDeadline(t *testing.T) {
+	sig := make(chan struct{})
+	var blockingWrite = func(conn net.Conn) {
+		conn.Write([]byte("0123456789"))
+		sig <- struct{}{}
+	}
+
+	var blockingRead = func(conn net.Conn) {
+		conn.Read(make([]byte, 10))
+		sig <- struct{}{}
+	}
+
+	p1, p2 := newPipe(5), newPipe(5)
+	c1, c2 := &conn{p1, p2}, &conn{p1, p2}
+
+	// Test with deadline
+	if err := c1.SetWriteDeadline(time.Now()); err != nil {
+		t.Fatalf("Error when setting deadline: %v", err)
+	}
+
+	go blockingWrite(c1)
+	select {
+	case <-time.After(100 * time.Millisecond):
+		t.Fatalf("Write timeout timed out, c = %v", c1)
+	case <-sig:
+	}
+
+	if err := c2.SetReadDeadline(time.Now()); err != nil {
+		t.Fatalf("Error when setting deadline: %v", err)
+	}
+
+	go blockingRead(c2)
+	select {
+	case <-time.After(100 * time.Millisecond):
+		t.Fatalf("Read timeout timed out, c = %v", c2)
+	case <-sig:
+	}
+
+	// Test without deadline
+	if err := c1.SetWriteDeadline(time.Time{}); err != nil {
+		t.Fatalf("Error when clearing deadline: %v", err)
+	}
+
+	if err := c2.SetReadDeadline(time.Time{}); err != nil {
+		t.Fatalf("Error when clearing deadline: %v", err)
+	}
+
+	go blockingWrite(c1)
+	select {
+	case <-time.After(100 * time.Millisecond):
+	case <-sig:
+		t.Fatal("Write returned before timeout")
+	}
+
```

Comment:
Optional: you could combine the final tests with this one: set a deadline here and expect an error.  Then you avoid the cleanup/reinitialization, and can make `c1`/`c2` cleanup via `defer`s.